### ISSUE TITLE
CSRF token with Angular

### ISSUE THREAD

**Issue by oddie** [adonisjs/adonis-framework#560]
Hi!

I'm setting up a new project with AdonisJS at backend and Angular 4 at frontend.
Stumbled with such a problem. AdonisJS sends a cookie XSRF-TOKEN in the response to a client. By default Angular adds this cookie in the next request at the same field named XSRF-TOKEN.
But AdonisJS only checks an input property of the request and the request headers.

```
/node_modules/adonis-middleware/src/Shield/index.js
...
 const csrfToken = request.input('_csrf') ||
    request.header('csrf-token') ||
    request.header('x-csrf-token') ||
    request.header('x-xsrf-token')
...
```
So then it returns an error "csrf token mismatch'.

This can be fixed this way

```
const csrfToken = request.input('_csrf') ||
    request.header('csrf-token') ||
    request.header('x-csrf-token') ||
    request.header('x-xsrf-token') ||
    request.plainCookie('XSRF-TOKEN')
```
Would you please add some additional checks over request cookies in the next release of AdonisJS? 

Thank you.

**Comment by thetutlage** [comment]
So I cannot see anything related to this on Angular4 docs. From initial angularJs docs. They say

> To take advantage of this, your server needs to set a token in a JavaScript readable session cookie called XSRF-TOKEN on the first HTTP GET request. On subsequent XHR requests the server can verify that the cookie matches X-XSRF-TOKEN HTTP header, and therefore be sure that only JavaScript running on your domain could have sent the request.

https://docs.angularjs.org/api/ng/service/$http#cross-site-request-forgery-xsrf-protection

So angular should send `X-XSRF-TOKEN` header

**Comment by oddie** [comment]
Hmm... This really works in AngularJs 1.5x. All headers are properly set, I've just checked. But not in Angular 2/4. I'd better ask them.

Thank you anyway.

**Comment by thetutlage** [comment]
@oddie Do update me if they think it is not a bug and they removed it intentionally, so I will add the support for same

**Comment by oddie** [comment]
@thetutlage It seems I found what's the problem. Angular2 do really send headers correctly if cookies are set and saved into the storage. It works out-of-the-box and doesn't need any additional configuration.

But it uses document.cookie to get all cookies saved into the browser's storage. So it can't read the cookies if they have "httpOnly" property set. At first you need to set httpOnly = false in session.js configuration file. 

```
/*
  |--------------------------------------------------------------------------
  | Http Only Cookie
  |--------------------------------------------------------------------------
  |
  | Keep cookie http only, which means javascript cannot access the cookie
  | by document.cookie.
  |
  */
  httpOnly: false,
...
```

Then I found that XSRF-TOKEN received within a response from the backend simply was not saved to a storage. Meanwhile cookies "adonis-session" and "adonis-session-value" was saved since they have all options "path", "domain" and "expires" set. XSRF-TOKEN has been wiped out without these options. I think for some security reasons.

So I slightly modified /node_modules/adonis-middleware/src/Shield/index.js to set missing options and all started working just fine.

```
// node_modules/adonis-middleware/src/Shield/index.js

class Shield {

  constructor (Config, View) {
    this.view = View
    this.csrf = new Csrf()
    this.shieldConfig = Config.get('shield')
    this.sessionConfig = Config.get('session') // <-- add this
    this.shieldConfig.csp.nonce = uuid.v4()
  }

......

* _setupCsrf (request, response, csrfSecret) {
    ...

    /**
     * adding token to the cookie to be used by JS frameworks
     * like Angular. Added a check for response.plainCookie
     * for backward compatibility, since plainCookie has
     * been added recently.
     */
    
    // add this
    const options = { 
      domain: this.sessionConfig.domain || null,
      path: this.sessionConfig.path || '/',
      secure: this.sessionConfig.secure || false,
      httpOnly: this.sessionConfig.httpOnly || false,
      sameSite: this.sessionConfig.sameSite || false
    }

    // remove this
    // typeof (response.plainCookie) === 'function'
    // ? response.plainCookie('XSRF-TOKEN', csrfToken)
    // : nodeCookie.create(request.request, response.response, 'XSRF-TOKEN', csrfToken)
    
    // add this
    nodeCookie.create(request.request, response.response, 'XSRF-TOKEN', csrfToken, options)
...


```

**Comment by thetutlage** [comment]
Cool! So I have made changes to the `adonis-middleware` package, would you mind using it from github and lemme know if it works fine. 

Now you can have custom cookie options for the CSRF token by defining them inside the config file which is https://github.com/adonisjs/adonis-app/blob/develop/config/shield.js#L87

Here's the fix https://github.com/adonisjs/adonis-middleware/commit/f28b5d116f8f9475a728ce34e3210dcb7d97fcd3

**Comment by oddie** [comment]
@thetutlage Now everything works fine. It's a good idea to make a separate config for CSRF token.
Thank you for the fast fix.

**Comment by thetutlage** [comment]
Cool, will release the patch in a while

**Comment by thetutlage** [comment]
Published as 1.0.14
